packages/dds/map/src/mapKernel.ts

```diff
@@ -168,6 +185,42 @@ export class MapKernel {
 	 */
 	private readonly localValueMaker: LocalValueMaker;
 
+	/**
+	 * The index to track the acknowledgement order of entries
+	 */
+	private insertionIndex: number = 0;
+
+	/**
+	 * Object to store the ack'd keys in creation order
+	 */
+	private readonly ackedKeysIndex: RedBlackTree<number, string> = new RedBlackTree(
+		compareNumbers,
+	);
+
+	/**
+	 * Object to store the mapping between the key and its insertion index
+	 */
+	private readonly ackedKeysTracker: Map<string, number> = new Map();
+
+	/**
+	 * Object to store the unack'd keys in creation order
+	 */
+	private readonly localKeysIndex: RedBlackTree<number, string> = new RedBlackTree(
+		compareNumbers,
+	);
+
+	/**
+	 * Object to store the message Id's for all set op's
+	 */
+	private readonly pendingSetTracker: Map<string, number[]> = new Map();
+
+	/**
+	 * Entries that have been deleted locally but not yet ack'd from the server. This maintains the record
+	 * of delete op that are pending or yet to be acked from server. This is maintained just to track the locally
+	 * deleted entries and the count of deleted times.
+	 */
+	private readonly pendingDeleteTracker: Map<string, number> = new Map();
```

Comment:
ah, interesting. I realize now that I'm not sure what the policy is here in a case like:

```
map1.set("foo", 5);
processAllMessages()
map1.delete("foo");
map2.set("foo", 4);
map1.delete("foo");
processAllMessages();
```

I would have expected the map contents to be {'foo' -> 4} as "map1 shouldn't be able to delete a key that it doesn't currently see". Is this the case? Or do we end up at the empty state? (I don't want to alter the merge outcome, but knowing the policy will be important)

Also--do we have test coverage for this merge outcome somewhere in existing map tests? I realize now that the comment I made earlier about `.delete()` happening on a key that exists being more interesting isn't really true. I think both scenarios are interesting now (so you should cover them both).

Reply:
I was confused about the scenario "should we permit a user to delete an entry coming from the future" as well, and I think it is not covered in the existing map tests. I suppose the overarching policy is we should avoid overwriting the value as long as there exists a local pending message targeting the same key, so the result should be an empty state. But I think it is worth a further discussion though..

Reply:
about to add this test in another PR